app/jobs/send_gpo_code_expiration_notices_job.rb

```diff
@@ -0,0 +1,77 @@
+class SendGpoCodeExpirationNoticesJob < ApplicationJob
+  queue_as :low
+
+  def initialize(analytics: nil)
+    @analytics = analytics
+  end
+
+  def perform
+    codes_to_send_notifications_for.find_each do |code|
+      user = code.profile.user
+
+      user.send_email_to_all_addresses(:gpo_code_expired, code_sent_at: code.code_sent_at)
+      code.update(expiration_notice_sent_at: Time.zone.now)
+
+      analytics.idv_gpo_expiration_email_sent(user_id: user.uuid)
+    end
+  end
+
+  def codes_to_send_notifications_for
+    # We are looking at a 48 hr window in which all codes will _definitely_ be expired.
+    from, to = calculate_notification_window_bounds
+    expired_codes_needing_notification_sent_between(
+      from: from,
+      to: to,
+    )
+  end
+
+  def calculate_notification_window_bounds(as_of: Time.zone.now)
+    to = as_of.beginning_of_day - IdentityConfig.store.usps_confirmation_max_days.days
+    from = to - 2.days
+    [from, to]
+  end
+
+  private
+
+  def analytics
+    @analytics ||= Analytics.new(user: AnonymousUser.new, request: nil, session: {}, sp: nil)
+  end
+
+  def expired_codes_needing_notification_sent_between(
+    from:,
+    to:
+  )
+    GpoConfirmationCode.joins(:profile).
+      # 1. Exclude codes that we've already sent an expiration notice for
+      where(expiration_notice_sent_at: nil).
+
+      # 2. Exclude codes not sent in the window we're looking at
```

Comment:
high level style q... I find all these spaces between these sections a little hard to read, and it kind of hides that its all one thing, what if we broke each scope into methods, and joined them? Then we could leave the long-form comments and they wouldn't interrupt the flow

```ruby
GpoConfirmationCode.joins(:profile).
  and(where_expiration_notice_nil).
  and(codes_sent_in_time_range(time_range)).
  and(profile_is_not_pending)
  # etc

 # 1. Exclude codes that we've already sent an expiration notice for
def where_expiration_notice_nil
  GpoConfirmationCode.where(expiration_notice_sent_at: nil)
end

# 2. Include only codes sent in the window we're looking at
def codes_sent_in_time_range(time_range)
  GpoConfirmationCode.where(code_sent_at: time_range).
end

# 3. Exclude codes where the associated profile does not have the GPO pending timestamp set
#    (meaning they either completed GPO or reset their password).
def profile_is_not_pending
  Profile.where(gpo_verification_pending_at: nil)
end

# etc...
```

Reply:
Yeah this is a good note. I broke the conditions out into methods and got rid of the comments altogether (since the method names say what they do)